### PR TITLE
fix: support URL and URLSearchParams

### PR DESCRIPTION
**Pull request by rturnq**
This PR adds support for URL and URLSearchParam objects

### CODE REVIEW

File: src/serialize.js [hunk 1]
```diff
@@ -38,6 +39,10 @@ function handleProperty(clone, key, value, valuePath, serializationSymbol, assig
         }
     } else if (value.constructor === Date) {
         assignments.push(valuePath + '=new Date(' + value.getTime() + ')');
+    } else if (value.constructor === URL) {
+        assignments.push(valuePath + '=new URL("' + value.href + '")');
```
Comment (DylanPiercey) [inline src/serialize.js:42]: I think this `href` will need to be escaped.

File: src/serialize.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 const constants = require("./constants");
+var { URLSearchParams } = require('url');
```
Comment (DylanPiercey) [inline src/serialize.js]: Since this file is loaded in the browser it should not import `url`.

File: src/serialize.js [hunk 3]
```diff
@@ -38,6 +39,10 @@ function handleProperty(clone, key, value, valuePath, serializationSymbol, assig
         }
     } else if (value.constructor === Date) {
         assignments.push(valuePath + '=new Date(' + value.getTime() + ')');
+    } else if (value.constructor === URL) {
+        assignments.push(valuePath + '=new URL("' + value.href + '")');
+    } else if (value.constructor === URLSearchParams) {
+        assignments.push(valuePath + '=new URLSearchParams("' + value.toString() + '")');
```
Comment (DylanPiercey) [inline src/serialize.js]: This should probably be escaped. Also I don't think the `toString` is necessary.

File: src/stringifyPrepare.js [hunk 4]
```diff
@@ -83,7 +88,7 @@ function pruneArray(array, path, serializationSymbol, assignments) {
 function pruneObject(obj, path, serializationSymbol, assignments) {
     var clone = {};
 
-    if (obj.toJSON && obj.constructor != Date) {
+    if (obj.toJSON && obj.constructor != Date && obj.constructor != URL && obj.constructor != URLSearchParams) {
```
Comment (DylanPiercey) [inline src/stringifyPrepare.js]: `URL` and `URLSearchParams` do not implement `toJSON`, so this change should not be needed.

**Review by DylanPiercey — CHANGES_REQUESTED** [review]

File: src/serialize.js (hunk 1 above)
Comment (rturnq) [reply src/serialize.js:42]: It appears that `"` is already url-encoded, should we replace any instances with `\"` to be sure?

File: src/stringifyPrepare.js (hunk 4 above)
Comment (rturnq) [reply src/stringifyPrepare.js]: Apparently URL does implement [toJSON](https://developer.mozilla.org/en-US/docs/Web/API/URL/toJSON) but I will remove it for URLSearchParams

File: src/serialize.js (hunk 1 above)
Comment (DylanPiercey) [reply src/serialize.js:42]: Good point. I think it should be fine then.

**Review by DylanPiercey — APPROVED** [review]

**Pull request merged**